prometheus/reporter.go

```diff
@@ -0,0 +1,269 @@
+func (r *reporter) RegisterGauge(name string, tags map[string]string, desc string) (*prom.GaugeVec, error) {
+	g := &prom.GaugeVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasGauge(id)
+	if exists {
+		return g, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+
+	g = prom.NewGaugeVec(
+		prom.GaugeOpts{
+			Name: name,
+			Help: desc,
+		},
+		labelKeys,
+	)
+	err := prom.Register(g)
+	if err != nil {
+		return g, err
+	}
+	r.Lock()
+	defer r.Unlock()
+	r.gauges[id] = g
+	return g, nil
+}
+
+// ReportGauge reports a gauge value
+func (r *reporter) ReportGauge(name string, tags map[string]string, value float64) {
+	id := hashMetricLabelsToID(name, tags)
+
+	r.RLock()
+	g, ok := r.gauges[id]
+	r.RUnlock()
+
+	if !ok {
+		var err error
+		g, err = r.RegisterGauge(name, tags, name+" gauge")
+		if err != nil {
+			panic(err)
+		}
+	}
+	g.With(tags).Set(value)
+}
+
+// RegisterTimer is a helper method to initialize a Timer histogram vector in the prometheus backend with a given help text.
+// If not called explicitly, the Reporter will create one for you on first use, with a not super helpful HELP string
+func (r *reporter) RegisterTimer(name string, tags map[string]string, desc string, objectives map[float64]float64) (*prom.SummaryVec, error) {
+	h := &prom.SummaryVec{}
+	id := hashMetricLabelsToID(name, tags)
+	exists := r.hasSummary(id)
+	if exists {
+		return h, errorAlreadyRegistered
+	}
+	labelKeys := keysFromMap(tags)
+
+	if objectives == nil {
+		objectives = r.objectives
+	}
+	h = prom.NewSummaryVec(
+		prom.SummaryOpts{
+			Name:       name,
+			Help:       desc,
+			Objectives: objectives,
+		},
+		labelKeys,
+	)
+	err := prom.Register(h)
+	if err != nil {
+		return h, err
+	}
+	r.Lock()
+	defer r.Unlock()
+	r.summaries[id] = h
+	return h, nil
+}
+
+// ReportTimer reports a timer value into the Summary histogram
+func (r *reporter) ReportTimer(name string, tags map[string]string, interval time.Duration) {
+	id := hashMetricLabelsToID(name, tags)
+
+	r.RLock()
+	h, ok := r.summaries[id]
+	r.RUnlock()
+
+	if !ok {
+		var err error
+		h, err = r.RegisterTimer(name, tags, name+" histogram in seconds", nil)
+		if err != nil {
+			panic(err)
+		}
+	}
+	h.With(tags).Observe(float64(interval))
+}
+
+// Capabilities ...
+func (r *reporter) Capabilities() tally.Capabilities {
+	return r
+}
+
+// Reporting indicates it can report outside of the process
+func (r *reporter) Reporting() bool {
+	return true
+}
+
+// Tagging indicates prometheus supports tagged metrics
+func (r *reporter) Tagging() bool {
+	return true
+}
+
+// Flush does nothing for prometheus
+func (r *reporter) Flush() {}
+
+// NOTE: this hashes name+label keys, not values, as we track metrics as Vectors, to support on-the-fly label changes
+func hashMetricLabelsToID(name string, tags map[string]string) metricID {
+	str := name + "{"
+	hasher := sha1.New()
```

Comment:
Sha1 is probably quite expensive and this function will be called a lot for timers which aren't buffered.  Might be best to just use the canonical ordered string as the key itself rather than hashing.